File: backend/dump.py
import socket
import struct

UDP_IP = "0.0.0.0"
UDP_PORT = 20777

sock = socket.socket(socket.AF_INET, socket.SOCK_DGRAM)
sock.bind((UDP_IP, UDP_PORT))
print("Listening for F1 25 telemetry...")

while True:
    data, addr = sock.recvfrom(2048)
    if len(data) < 29:
        continue
    
    header = struct.unpack_from('<HBBBBBQfIIBB', data, 0)
    packet_format = header[0]
    packet_id = header[5]
    player_car_index = header[10]
    
    if packet_id == 2:
        print(f"Packet Format: {packet_format}, Size: {len(data)}")
        
        # Try to find where the lap number is (usually 1, 2, 3...)
        # We'll dump the first 60 bytes of the player's lap data
        lap_size = (len(data) - 29) // 22
        print(f"Calculated lap_size: {lap_size}")
        
        offset = 29 + (player_car_index * lap_size)
        player_data = data[offset:offset+lap_size]
        
        print("Player Lap Data bytes:")
        for i in range(len(player_data)):
            print(f"{i}: {player_data[i]}")
            
        break

File: backend/dump.js
const dgram = require('dgram');
const server = dgram.createSocket('udp4');

server.on('message', (msg, rinfo) => {
  if (msg.length < 29) return;
  
  const packetFormat = msg.readUInt16LE(0);
  const packetId = msg.readUInt8(5);
  const playerCarIndex = msg.readUInt8(20); // wait, offset 20? Let's check header
  
  // Header:
  // 0: packetFormat (2)
  // 2: gameYear (1)
  // 3: gameMajorVersion (1)
  // 4: gameMinorVersion (1)
  // 5: packetVersion (1)
  // 6: packetId (1)
  // 7: sessionUID (8)
  // 15: sessionTime (4)
  // 19: frameIdentifier (4)
  // 23: overallFrameIdentifier (4)
  // 27: playerCarIndex (1)
  // 28: secondaryPlayerCarIndex (1)
  
  const pci = msg.readUInt8(27);
  
  if (packetId === 2) {
    console.log(`Format: ${packetFormat}, Length: ${msg.length}`);
    const lapSize = Math.floor((msg.length - 29) / 22);
    console.log(`Calculated Lap Size: ${lapSize}`);
    
    const offset = 29 + (pci * lapSize);
    console.log(`Player Car Index: ${pci}, Offset: ${offset}`);
    
    for (let i = 0; i < lapSize; i++) {
      console.log(`${i}: ${msg.readUInt8(offset + i)}`);
    }
    process.exit(0);
  }
});

server.bind(20777);
console.log("Listening on 20777...");
